packages/api/cms-api/src/user-permissions/auth/user-permissions.guard.ts

```diff
@@ -1,42 +1,65 @@
-import { CanActivate, ExecutionContext, Injectable } from "@nestjs/common";
+import { CanActivate, ExecutionContext, Inject, Injectable } from "@nestjs/common";
 import { Reflector } from "@nestjs/core";
-import { GqlExecutionContext } from "@nestjs/graphql";
+import { GqlContextType, GqlExecutionContext } from "@nestjs/graphql";
 
 import { CurrentUserInterface } from "../../auth/current-user/current-user";
 import { ContentScopeService } from "../content-scope.service";
-import { SCOPE_GUARD_ACTIVE_METADATA_KEY, ScopeGuardActiveMetadataValue } from "../decorators/scope-guard-active.decorator";
+import { RequiredPermission } from "../decorators/required-permission.decorator";
+import { ACCESS_CONTROL_SERVICE } from "../user-permissions.constants";
+import { AccessControlServiceInterface } from "../user-permissions.types";
 
 @Injectable()
 export class UserPermissionsGuard implements CanActivate {
-    constructor(protected reflector: Reflector, private readonly contentScopeService: ContentScopeService) {}
+    constructor(
+        protected reflector: Reflector,
+        private readonly contentScopeService: ContentScopeService,
+        @Inject(ACCESS_CONTROL_SERVICE) private readonly accessControlService: AccessControlServiceInterface,
+    ) {}
 
     async canActivate(context: ExecutionContext): Promise<boolean> {
         const isPublicApi = this.reflector.getAllAndOverride("publicApi", [context.getHandler(), context.getClass()]);
         if (isPublicApi) {
             return true;
         }
 
-        const scopeGuardActive = this.reflector.getAllAndOverride<ScopeGuardActiveMetadataValue | undefined>(SCOPE_GUARD_ACTIVE_METADATA_KEY, [
+        const request =
+            context.getType().toString() === "graphql" ? GqlExecutionContext.create(context).getContext().req : context.switchToHttp().getRequest();
+        const user = request.user as CurrentUserInterface | undefined;
+        if (!user) return false;
+
+        const requiredPermission = this.reflector.getAllAndOverride<RequiredPermission>("requiredPermission", [
             context.getHandler(),
             context.getClass(),
         ]);
+        if (!requiredPermission) {
+            throw new Error(`RequiredPermission decorator is missing in ${context.getClass().name}::${context.getHandler().name}()`);
+        }
 
-        if (scopeGuardActive === false) {
-            return true;
+        if (!this.isResolvingGraphQLField(context) && !requiredPermission.options?.skipScopeCheck) {
```

Comment:
So we don't check field resolvers for now? I think we should discuss if we want to check them in one of our meetings

Reply:
We do check field resolvers, but only the requiredPermission (see line 52) and since the ContentScope is definitely on a request-basis I think this is okay.

Reply:
You're right, I've missed that. Thought there was an early return.